tools/sof_perf_analyzer.py

```diff
@@ -0,0 +1,278 @@
+            trace_prev = trace_curr
+        trace_prev.timestamp += ts_shift
+        dispatch_trace_item(trace_prev)
+
+def process_kmsg_file():
+    '''Process the dmesg to get the component ID to component name mapping,
+    component name is acquired from the line that contains 'Create widget',
+    component ID is acquired from the next line. Example:
+    [  334.818435] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    Create widget host-copier.0.playback instance 0 - pipe 1 - core 0
+    [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]
+
+    In practice, sof-test only capture kernel message and firmware trace generated during a test
+    case run. Mostly in manual tests, if the kernel message file contains multiple firmware runs
+    with overlapping information, the last one wins.
+    '''
+    with open(args.kmsg, encoding='utf8') as f:
+        ppln_id = None
+        for line in f:
+            if match_obj := re.search(r'Create widget', line):
+                span_end_pos = match_obj.span()[1]
+                line_split = line[span_end_pos + 1:].split()
+                widget_name = line_split[0]
+                # In the linux kernel, IDA is used to allocated pipeline widget instance ID,
+                # this ID later is used for pipeline creation, thus becomes pipeline ID in the
+                # firmware. Note that ppln_id variable will be assigned properly at pipeline widget
+                # creation, because it is always the first one to be created before all other
+                # widgets in the same pipeline and pipelines are created sequentially.
+                if widget_name.startswith('pipeline'):
+                    ppln_id = line_split[2]
+                next_line = next(f)
+                widget_id = next_line.split('|')[0].split(':')[-1].strip()
+                # convert to the same ID format in mtrace
+                widget_id = '0x' + widget_id[:-6:-1]
+                component_name[Component(ppln_id, widget_id)] = widget_name
+
+def analyze_perf_info():
+    '''Calculate performance statistics from performance information'''
+    for comp, perf_info_list in perf_info.items():
+        len_perf_info = len(perf_info_list)
+        avg_cpu_avg = sum(e.cpu_avg for e in perf_info_list) / len_perf_info
+        avg_cpu_peak = sum(e.cpu_peak for e in perf_info_list) / len_perf_info
+        peak_cpu_avg = max(e.cpu_avg for e in perf_info_list)
+        peak_cpu_peak = max(e.cpu_peak for e in perf_info_list)
+        perf_stats[comp] = CompPerfStats(avg_cpu_avg, avg_cpu_peak, peak_cpu_avg, peak_cpu_peak)
+
+def print_perf_info():
+    '''Output SOF trace performance statistics'''
+    if len(perf_stats):
+        title_fmt = '{:>30}  {:10}  {:<15}  {:<15}  {:<15}  {:<15}'
+        title = title_fmt.format('COMP_NAME', 'COMP_ID', 'AVG_OF_CPU_AVG', 'AVG_OF_CPU_PEAK',
```

Comment:
If a separator character is added the output format would be CSV that eases import to spreadsheet. In US or C locale the separator is comma character. Could this be changed?

Reply:
sure, let's have this.